packages/legend-query/src/stores/QueryBuilderGraphFetchTreeUtil.ts

```diff
@@ -183,18 +186,46 @@ export const addQueryBuilderPropertyNode = (
   ];
   let parentExplorerTreeNode = explorerTreeData.nodes.get(node.parentId);
   while (
-    parentExplorerTreeNode instanceof QueryBuilderExplorerTreePropertyNodeData
+    parentExplorerTreeNode instanceof
+      QueryBuilderExplorerTreePropertyNodeData ||
+    parentExplorerTreeNode instanceof QueryBuilderExplorerTreeSubTypeNodeData
   ) {
-    const propertyGraphFetchTree = new PropertyGraphFetchTree(
-      PropertyExplicitReference.create(parentExplorerTreeNode.property),
-    );
-    propertyGraphFetchTree.subTrees.push(
-      propertyGraphFetchTrees[0] as PropertyGraphFetchTree,
-    );
-    propertyGraphFetchTrees.unshift(propertyGraphFetchTree);
-    parentExplorerTreeNode = explorerTreeData.nodes.get(
-      parentExplorerTreeNode.parentId,
-    );
+    let subType =
+      OptionalPackageableElementExplicitReference.create<Class>(undefined);
+    let subtypeAssigned = false;
+    while (
+      parentExplorerTreeNode instanceof QueryBuilderExplorerTreeSubTypeNodeData
+    ) {
+      if (!subtypeAssigned) {
+        subType = OptionalPackageableElementExplicitReference.create(
+          parentExplorerTreeNode.subclass,
+        );
+        subtypeAssigned = true;
+      }
+      parentExplorerTreeNode = explorerTreeData.nodes.get(
+        parentExplorerTreeNode.parentId,
+      );
+    }
+    if (
+      parentExplorerTreeNode instanceof QueryBuilderExplorerTreePropertyNodeData
+    ) {
+      const propertyGraphFetchTree = new PropertyGraphFetchTree(
+        PropertyExplicitReference.create(parentExplorerTreeNode.property),
+        subType,
+      );
+      propertyGraphFetchTree.subTrees.push(
+        propertyGraphFetchTrees[0] as PropertyGraphFetchTree,
+      );
+      propertyGraphFetchTrees.unshift(propertyGraphFetchTree);
+      parentExplorerTreeNode = explorerTreeData.nodes.get(
+        parentExplorerTreeNode.parentId,
+      );
+    } else {
+      queryBuilderState.applicationStore.notifyError(
+        `Can't cast the root class of graph fetch structure to its subtype. Only non-primitive properties of the root class could be cast to their subtypes in the graph fetch.`,
```

Comment:
Don't think we need the second sentence, the first one I think is succinct and the second just make it a bit more confusing.